File: frontend/src/types/clientVisit.ts
export type VisitType =
  | 'IN_PERSON_OFFICE'
  | 'CLIENT_SITE_VISIT'
  | 'VIRTUAL_CONFERENCE'
  | 'PRE_BID_MEETING'
  | 'COURTESY_CALL'
  | 'COMMERCIAL_NEGOTIATION'
  | 'OTHER';

export type VisitStatus =
  | 'SCHEDULED'
  | 'CHECKED_IN'
  | 'COMPLETED'
  | 'CANCELLED'
  | 'RESCHEDULED'
  | 'NO_SHOW';

export type VisitSentiment =
  | 'VERY_POSITIVE'
  | 'POSITIVE'
  | 'NEUTRAL'
  | 'CONCERNED'
  | 'CRITICAL';

export interface AccompanyingPerson {
  name: string;
  designation?: string;
  phone?: string;
  email?: string;
}

export interface ActionItem {
  task: string;
  owner?: string;
  deadline?: string;
  is_done: boolean;
}

export interface ClientVisit {
  id: string;
  title: string;
  client_organization: string;
  organization_id?: string | null;
  tender_id?: string | null;
  visitor_name: string;
  visitor_designation?: string | null;
  visitor_phone?: string | null;
  visitor_email?: string | null;
  accompanying_persons?: AccompanyingPerson[] | null;
  internal_host_name?: string | null;
  internal_host_role?: string | null;
  visit_type?: VisitType;
  status?: VisitStatus;
  scheduled_start: string;
  scheduled_end?: string | null;
  actual_check_in?: string | null;
  actual_check_out?: string | null;
  location_or_room?: string | null;
  meeting_link?: string | null;
  agenda?: string | null;
  discussion_notes?: string | null;
  action_items?: ActionItem[] | null;
  sentiment_outcome?: VisitSentiment | null;
  attachments?: Array<{ name: string; url?: string; size?: string }> | null;
  created_at?: string;
  updated_at?: string;
}

export interface ClientVisitCreateInput {
  title: string;
  client_organization: string;
  organization_id?: string | null;
  tender_id?: string | null;
  visitor_name: string;
  visitor_designation?: string | null;
  visitor_phone?: string | null;
  visitor_email?: string | null;
  accompanying_persons?: AccompanyingPerson[] | null;
  internal_host_name?: string | null;
  internal_host_role?: string | null;
  visit_type?: VisitType;
  status?: VisitStatus;
  scheduled_start: string;
  scheduled_end?: string | null;
  actual_check_in?: string | null;
  actual_check_out?: string | null;
  location_or_room?: string | null;
  meeting_link?: string | null;
  agenda?: string | null;
  discussion_notes?: string | null;
  action_items?: ActionItem[] | null;
  sentiment_outcome?: VisitSentiment | null;
}

File: backend/app/models/client_visit.py
from datetime import datetime, timezone
from sqlalchemy import Column, String, Text, DateTime, ForeignKey, JSON
from app.core.database import Base


class ClientVisit(Base):
    __tablename__ = "client_visits"

    id = Column(String(50), primary_key=True, index=True)
    title = Column(String(255), nullable=False)
    client_organization = Column(String(255), nullable=False, index=True)
    organization_id = Column(String(50), nullable=True, index=True)
    tender_id = Column(String(50), nullable=True, index=True)
    visitor_name = Column(String(150), nullable=False)
    visitor_designation = Column(String(150), nullable=True)
    visitor_phone = Column(String(50), nullable=True)
    visitor_email = Column(String(150), nullable=True)
    accompanying_persons = Column(JSON, nullable=True, default=list)
    internal_host_name = Column(String(150), nullable=True, default="Sarah Jenkins")
    internal_host_role = Column(String(100), nullable=True, default="Business Head")
    visit_type = Column(
        String(50),
        nullable=False,
        default="IN_PERSON_OFFICE",
        index=True,
    )  # IN_PERSON_OFFICE, CLIENT_SITE_VISIT, VIRTUAL_CONFERENCE, PRE_BID_MEETING, COURTESY_CALL, COMMERCIAL_NEGOTIATION
    status = Column(
        String(50),
        nullable=False,
        default="SCHEDULED",
        index=True,
    )  # SCHEDULED, CONFIRMED, CHECKED_IN, COMPLETED, RESCHEDULED, CANCELLED
    scheduled_start = Column(String(100), nullable=False)
    scheduled_end = Column(String(100), nullable=True)
    actual_check_in = Column(String(100), nullable=True)
    actual_check_out = Column(String(100), nullable=True)
    location_or_room = Column(String(150), nullable=True, default="Main Conference Room")
    meeting_link = Column(String(255), nullable=True)
    agenda = Column(Text, nullable=True)
    discussion_notes = Column(Text, nullable=True)
    action_items = Column(JSON, nullable=True, default=list)
    sentiment_outcome = Column(
        String(50), nullable=True, default="POSITIVE"
    )  # POSITIVE, NEUTRAL, CONCERNING, CRITICAL_BREAKTHROUGH
    attachments = Column(JSON, nullable=True, default=list)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )
